Can no longer build scipp with clang on Linux

Seems to have started happening around the time #404 was merged.

Clang version:
```
clang version 7.0.0-3~ubuntu0.18.04.1 (tags/RELEASE_700/final)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

Last output of `make -j VERBOSE=1`:
```
[ 76%] Linking CXX executable scipp-core-test
cd /home/dan/git/scipp/build/core/test && /usr/bin/cmake -E cmake_link_script CMakeFiles/scipp-core-test.dir/link.txt --verbose=1
ccache /usr/bin/clang++-7   -fno-omit-frame-pointer -fsized-deallocation -O3 -DNDEBUG   CMakeFiles/scipp-core-test.dir/bool_test.cpp.o CMakeFiles/scipp-core-test.dir/counts_test.cpp.o CMakeFiles/scipp-core-test.dir/dataset_test.cpp.o CMakeFiles/scipp-core-test.dir/dataset_test_common.cpp.o CMakeFiles/scipp-core-test.dir/dataset_operations_test.cpp.o CMakeFiles/scipp-core-test.dir/dimensions_test.cpp.o CMakeFiles/scipp-core-test.dir/except_test.cpp.o CMakeFiles/scipp-core-test.dir/transform_test.cpp.o CMakeFiles/scipp-core-test.dir/value_and_variance_test.cpp.o CMakeFiles/scipp-core-test.dir/variable_test.cpp.o CMakeFiles/scipp-core-test.dir/variable_custom_type_test.cpp.o CMakeFiles/scipp-core-test.dir/variable_operations_test.cpp.o CMakeFiles/scipp-core-test.dir/variable_view_test.cpp.o CMakeFiles/scipp-core-test.dir/view_index_test.cpp.o  -o scipp-core-test ../libscipp-core.a ../../googletest/lib/libgtest_main.a ../../googletest/lib/libgtest.a ../../units/libscipp-units.a ../../googletest/lib/libgtest.a -lpthread
../libscipp-core.a: error adding symbols: File format not recognised
clang: error: linker command failed with exit code 1 (use -v to see invocation)
core/test/CMakeFiles/scipp-core-test.dir/build.make:437: recipe for target 'core/test/scipp-core-test' failed
make[2]: *** [core/test/scipp-core-test] Error 1
make[2]: Leaving directory '/home/dan/git/scipp/build'
CMakeFiles/Makefile2:608: recipe for target 'core/test/CMakeFiles/scipp-core-test.dir/all' failed
make[1]: *** [core/test/CMakeFiles/scipp-core-test.dir/all] Error 2
make[1]: Leaving directory '/home/dan/git/scipp/build'
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -42,19 +42,6 @@ find_package(Boost 1.67 REQUIRED)
 
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-omit-frame-pointer")
 
-# See https://github.com/pybind/pybind11/issues/1604
-if(${CMAKE_CXX_COMPILER_ID} MATCHES "Clang")
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsized-deallocation")
-  if(APPLE)
-    add_compile_definitions(_APPLE_CLANG_)
-  endif()
-else()
-  set(CMAKE_CXX_VISIBILITY_PRESET hidden)
-
-  # Reduces binary size of, e.g., libscipp-core.so by several MB.
-  set(CMAKE_VISIBILITY_INLINES_HIDDEN TRUE)
-endif()
-
 option(WITH_CTEST "Enable ctest integration of tests" ON)
 option(DYNAMIC_LIB "Build shared libraries" OFF)
 
@@ -68,6 +55,24 @@ else ()
   set(CMAKE_INTERPROCEDURAL_OPTIMIZATION FALSE)
   message(WARNING "IPO is not supported: ${output}")
 endif (result)
+
+
+# See https://github.com/pybind/pybind11/issues/1604
+set(INTERPROCEDURAL_OPTIMIZATION_TESTS FALSE)
+if(${CMAKE_CXX_COMPILER_ID} MATCHES "Clang")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsized-deallocation")
+  if(APPLE)
+    add_compile_definitions(_APPLE_CLANG_)
+  else()
+    # For some reason clang6 on ubuntu18 can't link parts
+    # with different INTERPROCEDURAL_OPTIMIZATION flag.
+    set(INTERPROCEDURAL_OPTIMIZATION_TESTS ${CMAKE_INTERPROCEDURAL_OPTIMIZATION})
+  endif()
+else()
+  set(CMAKE_CXX_VISIBILITY_PRESET hidden)
+  # Reduces binary size of, e.g., libscipp-core.so by several MB.
+  set(CMAKE_VISIBILITY_INLINES_HIDDEN TRUE)
+endif()
 
 add_compile_options(-Wall
                     -Wextra

--- a/core/test/CMakeLists.txt
+++ b/core/test/CMakeLists.txt
@@ -27,7 +27,7 @@ target_link_libraries(${TARGET_NAME}
                       ${GMOCK_LIBRARIES})
 target_include_directories(${TARGET_NAME} SYSTEM
                            PRIVATE "../../range-v3/include")
-set_property(TARGET ${TARGET_NAME} PROPERTY INTERPROCEDURAL_OPTIMIZATION FALSE)
+set_property(TARGET ${TARGET_NAME} PROPERTY INTERPROCEDURAL_OPTIMIZATION ${INTERPROCEDURAL_OPTIMIZATION_TESTS})
 add_sanitizers(${TARGET_NAME})
 if(${WITH_CTEST})
   gtest_discover_tests(${TARGET_NAME} TEST_PREFIX scipp/core/)

--- a/neutron/test/CMakeLists.txt
+++ b/neutron/test/CMakeLists.txt
@@ -16,7 +16,7 @@ target_link_libraries(${TARGET_NAME}
                       scipp_test_helpers
                       ${GTEST_LIBRARIES}
                       ${GMOCK_LIBRARIES})
-set_property(TARGET ${TARGET_NAME} PROPERTY INTERPROCEDURAL_OPTIMIZATION FALSE)
+set_property(TARGET ${TARGET_NAME} PROPERTY INTERPROCEDURAL_OPTIMIZATION ${INTERPROCEDURAL_OPTIMIZATION_TESTS})
 add_sanitizers(${TARGET_NAME})
 if(${WITH_CTEST})
   gtest_discover_tests(${TARGET_NAME} TEST_PREFIX scipp/neutron/)

--- a/units/test/CMakeLists.txt
+++ b/units/test/CMakeLists.txt
@@ -9,7 +9,7 @@ target_link_libraries(${TARGET_NAME}
                       scipp_test_helpers
                       ${GTEST_LIBRARIES}
                       ${GMOCK_LIBRARIES})
-set_property(TARGET ${TARGET_NAME} PROPERTY INTERPROCEDURAL_OPTIMIZATION FALSE)
+set_property(TARGET ${TARGET_NAME} PROPERTY INTERPROCEDURAL_OPTIMIZATION ${INTERPROCEDURAL_OPTIMIZATION_TESTS})
 add_sanitizers(${TARGET_NAME})
 if(${WITH_CTEST})
   gtest_discover_tests(${TARGET_NAME} TEST_PREFIX scipp/units/)

